### PR TITLE
Do chapter 2

### PR DESCRIPTION
**Pull request by cherealnice**
### Solutions for Chapter 2

cc @vrom911 @chshersh

### CODE REVIEW

File: src/Chapter2.hs [hunk 1]
```diff
@@ -783,10 +812,12 @@ divideTenBy :: Int -> Int
...
 
 -- TODO: type ;)
-listElementsLessThan x l = filter (< x) l
+listElementsLessThan :: Ord a => a -> [a] -> [a]
+listElementsLessThan x = filter (< x)
 
 -- Can you eta-reduce this one???
-pairMul xs ys = zipWith (*) xs ys
+pairMul :: [Integer] -> [Integer] -> [Integer]
+pairMul = zipWith (*)
```
Comment (vrom911) [inline src/Chapter2.hs:820]: Eta-reduction award 🏆

File: src/Chapter2.hs [hunk 2]
```diff
@@ -841,7 +872,8 @@ list.
 
 🕯 HINT: Use the 'cycle' function
 -}
-rotate = error "rotate: Not implemented!"
+rotate :: Int -> [a] -> [a]
+rotate offset list = subList offset (length list + (offset - 1)) $ cycle list
```
Comment (vrom911) [inline src/Chapter2.hs:876]: Interesting solution with `subList`! 💯 

Unfortunately, `cycle` fails in runtime on empty lists :recycle: So you need to handle this case separately for this function to work properly.

Also, even hat's a correct solution, though, it does a little bit more than work when the given number is big and the list is small. You will cycle through the list a lot. Maybe you can think of a way to optimize this :slightly_smiling_face:

File: src/Chapter2.hs [hunk 3]
```diff
@@ -857,7 +889,8 @@ and reverses it.
   function, but in this task, you need to implement it manually. No
   cheating!
 -}
-rewind = error "rewind: Not Implemented!"
+rewind :: [a] -> [a]
+rewind list = map (list !!) [length list - 1, length list - 2..0]
```
Comment (vrom911) [inline src/Chapter2.hs:893]: Creative solution! 🎨 

Though it is slow as `!!`, `length` walks each element of the list to get its size, so on each step, you will go through all the elements, which will take a lot! And imagine making both of them for each element at every step.
Also `!!` is not a safe function and can fail in runtime if the function will be outside the length of the list.

That is why it is always better to rewrite it with the `:` cons. Remember the explanation with trains? 🚂 🚋 🚋 

That is why a more efficient solution is with the accumulator and the recursive function that will do the addition at the start of the list which is instant!

You can read a bit more about the `go` pattern in here: https://kowainik.github.io/posts/haskell-mini-patterns#recursive-go

**Review by vrom911 — APPROVED** [review]
Great job! Chapter Two is done ✔️ Great work 👏🏼

File: src/Chapter2.hs [hunk 4]
```diff
@@ -500,7 +521,9 @@ True
 >>> isThird42 [42, 42, 0, 42]
 False
 -}
-isThird42 = error "isThird42: Not implemented!"
+isThird42 :: (Eq a, Num a) => [a] -> Bool
+isThird42 (_:_:42:_) = True
+isThird42 _ = False;
```
[the remark this answers is not in the document]
Comment (cherealnice) [reply src/Chapter2.hs:526]: Oops! Habit...

File: src/Chapter2.hs [hunk 5]
```diff
@@ -620,7 +643,13 @@ Write a function that takes elements of a list only on even positions.
...
 [2,3,4]
 -}
-takeEven = error "takeEven: Not implemented!"
+takeEven :: [a] -> [a]
+takeEven = go
+  where
+    go :: [a] -> [a]
+    go [] = []
+    go [x] = [x]
+    go (x:_:xs) = x : takeEven xs
```
[the remark this answers is not in the document]
Comment (cherealnice) [reply src/Chapter2.hs:652]: Yea I wanted to use `go` pattern here, but maybe it wasn't necessary :)

File: src/Chapter2.hs (hunk 3 above)
Comment (cherealnice) [reply src/Chapter2.hs]: Thanks for the suggestions! I took some more time for this one, and might have cheated a bit but I understand it a lot better now :D

File: src/Chapter2.hs (hunk 2 above)
Comment (cherealnice) [reply src/Chapter2.hs]: Thanks for the suggestion, I took another shot at it

**Comment by vrom911** [comment]
The solutions for this Chapter look awesome 😍 
I'm going to close this PR, because we can't merge the solutions to our repo :slightly_smiling_face: But don't worry, it still counts towards Hacktoberfest contribution :shirt: :deciduous_tree: Feel free to open PRs with solutions to other chapters!

**Pull request closed without merging**